flocker/node/_loop.py

```diff
@@ -300,10 +307,54 @@ def __init__(self, reactor, deployer):
         self.deployer = deployer
         self.cluster_state = None
 
+        # Save last known local state, and AMP client we sent it to.
+        # Since only one control service can exist at the moment, there can
+        # no more than one AMP client at any given point in time.
+        self.lock = threading.Lock()
+        self.last_sent_local_state = None
+        self.last_sent_client = None
+
     def output_STORE_INFO(self, context):
         self.client, self.configuration, self.cluster_state = (
             context.client, context.configuration, context.state)
 
+    def _maybe_send_state_to_control_service(self, state_changes):
+        """
+        If current state_changes differ from last sent local state,
+        transmit the new state to control agent.
+
+        :param state_changes: State to send to the control service.
+        :type state_changes: tuple of IClusterStateChange
+        """
+        if (self.last_sent_local_state != state_changes or
+           self.last_sent_client != self.client):
+            context = LOG_SEND_TO_CONTROL_SERVICE(
+                self.fsm.logger, connection=self.client,
+                local_changes=list(state_changes),
+            )
+            with context.context():
+                d = DeferredContext(self.client.callRemote(
+                    NodeStateCommand,
+                    state_changes=state_changes,
+                    eliot_context=context)
+                )
+
+                def set_sent_state(_):
+                    with self.lock:
+                        self.last_sent_local_state = state_changes
+                        self.last_sent_client = self.client
+
+                def clear_sent_state(f):
+                    with self.lock:
+                        self.last_sent_local_state = None
+                        self.last_sent_client = None
+                    return f
+                d.addCallbacks(set_sent_state, clear_sent_state)
+                d.addErrback(
+                    writeFailure, self.fsm.logger,
+                    u"Failed to send local state to control node.")
```

Comment:
Convergence loop seems to have specific logger (https://github.com/ClusterHQ/flocker/blob/master/flocker/node/_loop.py#L418), and associated tests (https://github.com/ClusterHQ/flocker/blob/master/flocker/node/test/test_loop.py#L384). Removing logger parameter from above line regressed one of the 37 tests:

```
Traceback (most recent call last):
  File "/home/ubuntu/flocker/.tox/py27/local/lib/python2.7/site-packages/twisted/trial/_synctest.py", line 1368, in _runCleanups
    f(*args, **kwargs)
  File "/home/ubuntu/flocker/.tox/py27/local/lib/python2.7/site-packages/eliot/testing.py", line 274, in <lambda>
    self, logger, *assertionArgs, **assertionKwargs))
  File "/home/ubuntu/flocker/flocker/node/test/test_loop.py", line 559, in assert_full_logging
    u"desired_configuration": self.configuration})
  File "/home/ubuntu/flocker/.tox/py27/local/lib/python2.7/site-packages/eliot/testing.py", line 385, in assertHasAction
    testCase.assertTrue(actions, "No actions of type %s" % (actionType,))
  File "/home/ubuntu/flocker/.tox/py27/local/lib/python2.7/site-packages/twisted/trial/_synctest.py", line 403, in assertTrue
    raise self.failureException(msg)
twisted.trial.unittest.FailTest: No actions of type <eliot._validation.ActionType object at 0x7f49e9387c90>

flocker.node.test.test_loop.ConvergenceLoopFSMTests.test_convergence_done_delays_new_iteration

```

In the interest of not refactoring this code, am leaving the ``logger`` parameter as is. Please let me know if you think this change needs to be checked in with current PR. Thanks!